src/HEVM.hs

```diff
@@ -121,12 +121,12 @@ checkPostStorage ctx (Behaviour _ _ _ _ _ _ updates _) pre post contractMap solc
               (Concrete pre', Concrete post') -> literal $ pre' == post'
               (Symbolic _ pre', Symbolic _ post') ->
                let
-                 insertions = rights $ filter (\a -> addr == get (getContractId (getLoc a)) contractMap) updates
-                 slot update' = let S _ w = calculateSlot ctx solcjson (mkLoc update')
+                 insertions = rights $ filter (\a -> addr == get (contractFromRewrite a) contractMap) updates
+                 slot update' = let S _ w = calculateSlot ctx solcjson (locFromUpdate update')
                                in w
                  insertUpdate :: SArray (WordN 256) (WordN 256) -> StorageUpdate -> SArray (WordN 256) (WordN 256)
-                 insertUpdate store u@(IntUpdate _ e) = writeArray store (slot u) $ sFromIntegral $ symExpInt ctx Pre e
-                 insertUpdate store u@(BoolUpdate _ e) = writeArray store (slot u) $ ite (symExpBool ctx Pre e) 1 0
+                 insertUpdate store u@(IntUpdate _ e) = writeArray store (slot u) $ sFromIntegral $ symExpInt ctx $ e `as` Pre
```

Comment:
What does this `as` mean?

Reply:
That we are taking an expression which has no explicit timing (i.e. `Untimed`) and points it towards the prestate. `as` is a method of the class `Timeable`, but in this particular case it will be

`as :: Exp Untimed a -> When -> Exp Timed a`.

The point of this is that within a `Behaviour` or a `Constructor`, some expressions will have an explicit timing and some will not, but we still want the variable names that the backend generates to match between them. So most backends will want to require that its input is *always* `Timed` (requiring `Untimed` instead is fine only if the backend doesn't care about postconditions and return values), which `as` gives us a way to supply.

The `Timeable` class also has `setPre :: Exp Untimed a -> Exp Timed a` which is literally the same as `flip as Pre` (and of course `setPost` too) and `setTime :: When -> Exp Untimed a -> Exp Timed a` which is just`flip as`. We may want to decide on one particular style and remove the other methods?

Reply:
No longer relevant due to redesign, closing.